app/src/main/java/ch/eureka/eurekapp/model/data/notes/UnifiedSelfNotesRepository.kt

```diff
@@ -0,0 +1,182 @@
+
+          val messageWithUser = message.copy(senderId = userId)
+          val isCloudEnabled = userPreferences.isCloudStorageEnabled.first()
+
+          val entity =
+              messageWithUser
+                  .toEntity()
+                  .copy(isPendingSync = isCloudEnabled, isPrivacyLocalOnly = !isCloudEnabled)
+
+          // Always save locally first (Offline capability)
+          localDao.insertMessage(entity)
+
+          // Only attempt Cloud upload if Cloud is enabled AND we are online.
+          if (isCloudEnabled) {
+            if (isInternetAvailable()) {
+              try {
+                val result = firestoreRepo.createNote(messageWithUser)
+                if (result.isSuccess) {
+                  localDao.markAsSynced(message.messageID)
+                }
+              } catch (e: Exception) {
+                Log.e("UnifiedRepo", "Immediate upload failed, will sync later via Worker", e)
+              }
+            } else {
+              Log.d("UnifiedRepo", "Offline: Note stored locally, pending sync.")
+            }
+          }
+
+          Result.success(message.messageID)
+        } catch (e: Exception) {
+          Result.failure(e)
+        }
+      }
+
+  override suspend fun deleteNote(noteId: String): Result<Unit> =
+      withContext(Dispatchers.IO) {
+        val userId = getCurrentUserId()
+        localDao.deleteMessage(noteId, userId)
+        // Only try to delete from cloud if online
+        if (isInternetAvailable()) {
+          firestoreRepo.deleteNote(noteId)
+        }
+        Result.success(Unit)
+      }
+
+  /**
+   * Updates the user's preference for storage location (Local-only vs. Cloud).
+   *
+   * This function handles the logic for switching modes:
+   * - Saves the new preference to DataStore.
+   * - If switching TO Cloud mode (`enableCloud = true`), it finds all existing local notes that
+   *   were previously marked as "private/local-only", marks them as pending sync, and immediately
+   *   attempts to upload them.
+   *
+   * This method is main-safe as it wraps blocking database calls in [Dispatchers.IO].
+   *
+   * @param enableCloud True to enable Cloud storage and sync; False to revert to Local-only
+   *   storage.
+   * @return The number of notes successfully synced to the cloud during this operation (0 if
+   *   disabling cloud).
+   */
+  suspend fun setStorageMode(enableCloud: Boolean): Int =
+      withContext(Dispatchers.IO) {
+        userPreferences.setCloudStorageEnabled(enableCloud)
+
+        var syncedCount = 0
+        if (enableCloud) {
+          val userId = getCurrentUserId()
+          // If switching to Cloud, mark all existing local private notes as "Ready for Sync"
+          localDao.makeAllMessagesPublicForUser(userId)
+          // Trigger a sync and return the count
+          if (isInternetAvailable()) {
+            syncedCount = syncPendingNotes()
+          }
```

Comment:
Not atomic and it will leave data in an inconsistent state. If the app crashes between these lines (and apps do crash...), all the user's supposedly private notes are now marked as public but never actually synced to the cloud. They've lost their privacy guarantee with no benefit. Stop trying to do this synchronously. Use WorkManager properly to handle the entire operation atomically:

```suggestion
suspend fun setStorageMode(enableCloud: Boolean) {
    userPreferences.setCloudStorageEnabled(enableCloud)
    if (enableCloud) {
      scheduleStorageModeTransitionWorker()
    }
  }
```
You may create a dedicated worker that marks notes as public AND syncs them in a single atomic operation. If it fails, the whole thing fails and can be retried properly.